joinservice/internal/certcache/certcache.go

```diff
@@ -0,0 +1,165 @@
+*/
+
+// Package certcache implements an in-cluster SEV-SNP certificate cache.
+package certcache
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"strings"
+
+	"github.com/edgelesssys/constellation/v2/internal/attestation/variant"
+	"github.com/edgelesssys/constellation/v2/internal/constants"
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/edgelesssys/constellation/v2/joinservice/internal/certcache/amdkds"
+	"github.com/edgelesssys/constellation/v2/joinservice/internal/kubernetes"
+	"github.com/google/go-sev-guest/abi"
+	"github.com/google/go-sev-guest/verify/trust"
+)
+
+// Client is a client for interacting with the certificate chain cache.
+type Client struct {
+	log        *logger.Logger
+	attVariant *variant.Variant
+	kdsClient
+	kubeClient kubeClient
+}
+
+// NewClient creates a new CertCacheClient.
+func NewClient(log *logger.Logger, kubeClient kubeClient, attVariant *variant.Variant) *Client {
+	kdsClient := amdkds.NewKDSClient(trust.DefaultHTTPSGetter())
+
+	return &Client{
+		attVariant: attVariant,
+		log:        log,
+		kubeClient: kubeClient,
+		kdsClient:  kdsClient,
+	}
+}
+
+// CreateCertChainCache creates a certificate chain cache for the given attestation variant
+// and returns the cached certificates, if applicable.
+// If the certificate chain cache already exists, nothing is done.
+func (c *Client) CreateCertChainCache(ctx context.Context) (*CachedCerts, error) {
+	switch *c.attVariant {
+	case variant.AzureSEVSNP{}:
+		c.log.Debugf("Creating azure SEV-SNP certificate chain cache")
+		ask, ark, err := c.createCertChainCache(ctx, abi.VcekReportSigner)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create azure SEV-SNP certificate chain cache: %w", err)
+		}
+		return &CachedCerts{
+			ask: ask,
+			ark: ark,
+		}, nil
+	// TODO(derpsteb): Add AWS
+	default:
+		c.log.Debugf("No certificate chain caching possible for attestation variant %s", *c.attVariant)
+		return nil, nil
+	}
+}
+
+// CachedCerts contains the cached certificates.
+type CachedCerts struct {
+	ask *x509.Certificate
+	ark *x509.Certificate
+}
+
+// SevSnpCerts returns the cached SEV-SNP ASK and ARK certificates.
+func (c *CachedCerts) SevSnpCerts() (ask, ark *x509.Certificate) {
+	return c.ask, c.ark
+}
+
+// createCertChainCache creates a certificate chain cache configmap with the ASK and ARK
+// retrieved from the KDS and returns ASK and ARK.
+// If the configmap already exists, nothing is done and the existing ASK and ARK are returned.
+func (c *Client) createCertChainCache(ctx context.Context, signingType abi.ReportSigner) (ask, ark *x509.Certificate, err error) {
+	c.log.Debugf("Creating certificate chain cache")
+	ask, ark, err = c.getCertChainCache(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get certificate chain cache: %w", err)
+	}
+	if ask != nil && ark != nil {
+		c.log.Debugf("Certificate chain cache already exists")
+		return ask, ark, nil
+	}
```

Comment:
As said in my previous comment, we should check for the error of `getCertChainCache` to decide whether the cert cache ConfigMap already exists.
However, we might want to implement additional logic to update an existing cert cache ConfigMap in case the ASK or ARK are missing from it.
In this case we would fetch the missing certificate and then call an Update function instead of a Create function.